File: 2904-shortest-and-lexicographically-smallest-beautiful-string/2904-shortest-and-lexicographically-smallest-beautiful-string.go
func shortestBeautifulSubstring(s string, k int) string {
	ss := ""
	count := 0
	left := 0
	for right := 0; right < len(s); right++ {
		if s[right] == '1' {
			count++
		}

		if count == k {
			for s[left] == '0' {
				left++
			}

			curr := s[left : right+1]
			if ss == "" || len(curr) < len(ss) || (len(curr) == len(ss)) && (curr < ss) {
				ss = curr
			}

            // discard the first '1' so that the next window looks for the next k '1's
            left++
            count--
		}
	}

	return ss
}

// sliding window, string
// time: O(n)
// space: O(1)

File: 2904-shortest-and-lexicographically-smallest-beautiful-string/2904-shortest-and-lexicographically-smallest-beautiful-string.py
class Solution:
    def shortestBeautifulSubstring(self, s: str, k: int) -> str:
        ss = ""
        left = count = 0
        for right in range(len(s)):
            if s[right] == '1':
                count += 1
            
            if count == k:
                while s[left] == '0':
                    left += 1
                
                curr = s[left:right+1]
                if ss == '' or len(curr) < len(ss) or (len(curr) == len(ss) and curr < ss):
                    ss = curr

                left += 1
                count -= 1
        
        return ss

# sliding window, string
# time: O(n)
# space: O(1)